nexus/reconfigurator/simulation/src/system.rs

```diff
@@ -0,0 +1,473 @@
+    //    `Arc<Mutex<Store>>`. Our current approach is more simplistic, but
+    //    also lock-free.
+    //
+    // None of these are insurmountable, but they do make the global store a
+    // bit less appealing than it might seem at first glance.
+    //
+    /// Describes the sleds in the system.
+    ///
+    /// This resembles what we get from the `sled` table in a real system.  It
+    /// also contains enough information to generate inventory collections that
+    /// describe the system.
+    description: SystemDescription,
+
+    /// Inventory collections created by the user.
+    ///
+    /// Stored with `Arc` to allow cheap cloning.
+    collections: IndexMap<CollectionUuid, Arc<Collection>>,
+
+    /// Blueprints created by the user.
+    ///
+    /// Stored with `Arc` to allow cheap cloning.
+    blueprints: IndexMap<Uuid, Arc<Blueprint>>,
+
+    /// Internal DNS configurations.
+    ///
+    /// Stored with `Arc` to allow cheap cloning.
+    internal_dns: BTreeMap<Generation, Arc<DnsConfigParams>>,
+
+    /// External DNS configurations.
+    ///
+    /// Stored with `Arc` to allow cheap cloning.
+    external_dns: BTreeMap<Generation, Arc<DnsConfigParams>>,
+}
+
+impl SimSystem {
+    pub fn new() -> Self {
+        Self {
+            description: SystemDescription::new(),
+            collections: IndexMap::new(),
+            blueprints: IndexMap::new(),
+            internal_dns: BTreeMap::new(),
+            external_dns: BTreeMap::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        !self.description.has_sleds()
+            && self.collections.is_empty()
+            && self.blueprints.is_empty()
+            && self.internal_dns.is_empty()
+            && self.external_dns.is_empty()
+    }
+
+    #[inline]
+    pub fn description(&self) -> &SystemDescription {
+        &self.description
+    }
+
+    pub fn get_collection(
+        &self,
+        id: CollectionUuid,
+    ) -> Result<&Collection, KeyError> {
+        match self.collections.get(&id) {
+            Some(c) => Ok(&**c),
+            None => Err(KeyError::collection(id)),
+        }
+    }
+
+    pub fn all_collections(
+        &self,
+    ) -> impl ExactSizeIterator<Item = &Collection> {
+        self.collections.values().map(|c| &**c)
+    }
+
+    pub fn get_blueprint(&self, id: Uuid) -> Result<&Blueprint, KeyError> {
+        match self.blueprints.get(&id) {
+            Some(b) => Ok(&**b),
+            None => Err(KeyError::blueprint(id)),
+        }
+    }
+
+    pub fn all_blueprints(&self) -> impl ExactSizeIterator<Item = &Blueprint> {
+        self.blueprints.values().map(|b| &**b)
+    }
+
+    pub fn get_internal_dns(
+        &self,
+        generation: Generation,
+    ) -> Result<&DnsConfigParams, KeyError> {
+        self.internal_dns
+            .get(&generation)
+            .map(|d| &**d)
+            .ok_or_else(|| KeyError::internal_dns(generation))
```

Comment:
I don't really have a preference, but `get_{collection,blueprint}` use a `match` and `get_{internal,external}_dns` use `.map().ok_or_else()` to do the same thing, I think? Probably would be less surprising if they all used the same style.

Reply:
Agreed, changed everything to `match`.